sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+        the edges fanning out of the root nodes. Every node is visited once; for each node, every outgoing synapse
+        is forwarded. That way, we are guaranteed to forward any component exactly once on every simulation step.
+
+        Parameters
+        ----------
+        data: torch.tensor
+            External input to be processed by this generic network.
+
+        Returns
+        -------
+        dict
+            Dictionary containing the loggable properties of this network's ensembles and synapses in this
+            simulation step.
+
+        Warning
+        -------
+        Networks now support multiple root nodes and external input streams. In those cases, the graph is
+        traversed layer-wise, using multi-source BFS.
+
+        See Also
+        --------
+        :meth:`simulation.Simulator.run`
+            For a demonstration of how to programmatically feed heterogeneous external current to multiple nodes.
+            Note that this functionality will eventually be relegated to the data loader, with arbitrary current
+            injections during the simulation being implemented by biases.
+
+        """
+        # follow outgoing synapses between vertices, forwarding the entire network on each iteration.
+        for i, ensemble in enumerate(self.traversal_order):
+            # list all synapses coming in and out of this ensemble. Networkx references it by its string identifier.
+            incoming_synapses = self._in_edges(ensemble)
+            outgoing_synapses = self._out_edges(ensemble)
+
+            # shortcut to actual ensemble reference.
+            ensemble_ref = self.graph.nodes[ensemble].get("reference")
+
+            if ensemble_ref.identifier not in self.roots:
+                # apply a summation function to synaptic data flowing into this ensemble (torch.sum by default).
+                integrated_data = self.summation([synapse.output for synapse in incoming_synapses]).to(self.device)
+
+            else:
+                # if this is a root node, treat its stream from the data loader as inbound synaptic input.
+                external = [data[self.roots.index(ensemble_ref.identifier)]] if isinstance(data, list) else [data]
+                feedback = [synapse.output for synapse in incoming_synapses]
+
+                integrated_data = self.summation(external + feedback)
+
+            # forward current ensemble.
+            ensemble_ref(integrated_data)
+
+            # use current ensemble's analog or spike data to forward its outgoing synapses.
+            if hasattr(ensemble_ref, "spiked"):
+                [synapse(ensemble_ref.spiked.float()) for synapse in outgoing_synapses]
+
+            elif hasattr(ensemble_ref, "voltage"):
+                [synapse(ensemble_ref.voltage) for synapse in outgoing_synapses]
+
+        self.simulation_step += 1
+
+        # build and pass a state dictionary for use by calling module.
+        state = {"ensembles": {}, "synapses": {}}
```

Comment:
I know this is probably a large question but why don't we use hooks for everything? Also, would there potentially be a difference between getting the logs as we pass data through neurons/synapses vs. at the end? Does each neuron/synapse only get data passed through it once and only once for each forward chain?

Reply:
Yes, every node or edge is visited exactly once on every Network.forward call. For nodes, this is guaranteed by BFS. Then, since all synapses have exactly one source and the algorithm only forwards outgoing edges (at each node), you can't visit an edge twice and you can't not visit it at all.

Other than streaming intermediate states to disk, what else would you use forward hooks for?

Reply:
>Other than streaming intermediate states to disk, what else would you use forward hooks for?

Well, it saves having to pass state around!? And users can perhaps decide which states to save in the hook, rather than always getting and passing the state around so that you're always paying the cost for everything. By using a forward hook, you leave it all to the hook and you can use different hooks on the user side, unlike passing the loggables around, which you do at the framework level and users have no control over.

This is part of the convo about Loggable and Configurable. I think how you saw loggable is different than what I had initially intended it as. As I had originally intended, unlike Configurable where the property value is stored in the object, loggable doesn't store the exact properties to be logged as that was stored in a separate dict. That is, like configurable, `_loggable_props_` was meant as a hard list of all the potential props that can be logged. However, to decide at runtime which props are actually logged, that's what I had a yaml file for, where each potential loggable property is marked whether to be logged or not. And, during runtime, the dicts created from the yaml can be dynamically updated to change some loggables from True to False etc.

Your interpretation instead seems to be that everything listed in loggable_props are to be logged. And if we need to change that, then we need to update loggable_props. Vs. updating a separate dict and keeping loggable_props unchanged.

This is where some of that tension of how you are using loggable_props and extensible_props is coming from I think.

Anyway, that's all to say, if you did follow my approach, then blindly saving all loggables after every forward at a framework level may be wasteful as users may not want to log everything listed in loggable. And using customizable forward hooks may be more flexible.

On the other hand, users can dynamically override so that `loggable_state()` returns different things dynamically, maybe based on a filter dict changing which props are logged for each forward call, so perhaps this is not an issue. This is just something for you to consider.

Reply:
> Your interpretation instead seems to be that everything listed in loggable_props are to be logged. And if we need to change that, then we need to update loggable_props. Vs. updating a separate dict and keeping loggable_props unchanged.

Exactly. This rhymes with your preference for class attributes and mine for instance attributes.

All good points. I anticipate a broad solution to the loggable/configurable conundrum involving a parent class that would attach descriptors/tags to properties (for lack of a better term; I wouldn't use a python-reserved word), specifying what's configurable/loggable/extensible etc. (whether possibly or in actuality) and implementing shared functionality across your classes. In the abstract, any attribute can have multiple appropriate uses at runtime. It can wait until we have some downtime or more hands on deck.